### ISSUE TITLE
feat: Actual type hints (not representative) for UDFs

### ISSUE THREAD

**Issue by NickCrews** [ibis-project/ibis#7201]
### Is your feature request related to a problem?

Sorry that I'm making these issues on UDFs when I'd imagine you are about to release them :(

This is what I currently do:

```python
import pyarrow.compute as pc

@ibis.udf.scalar.pyarrow
def add_one(x: int) -> int:
    return pc.add(x, 1)
```

In the function, `x` isn't a python int, but is actually a pyarrow chunked array of type int. This both is confusing to me as the reader of this function, and it makes my editor sad.

### Describe the solution you'd like

Could we annotate in such a way that standard python tooling eg mypy is happy? Maybe using typing.Annotated

```python
from typing import Annotated

```python
import ibis
from pyarrow.lib import IntegerArray
from typing import Annotated
import pyarrow.compute as pc


@ibis.udf.scalar.pyarrow
def add_one(x: Annotated[IntegerArray, int]) -> Annotated[IntegerArray, int]:
    return pc.add(x, 1)

t = ibis.examples.penguins.fetch()
result = add_one(t.year)
```

### What version of ibis are you running?

main

### What backend(s) are you using, if any?

NA

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cpcloud** [comment]
UDF type hints are meant to be convenient for describing the element type of the inputs, not comprehensively compatible with mypy.

Is there some way we can have both?

I suppose we could add support for the `Annotated` versions of things for people who want better mypy compatibility.

I think this would ship after 7.0.

**Comment by NickCrews** [comment]
I think accepting either `int` or `Annotated[pyarrow_type, int]` would just be extra code, but should work fine?

Thinking about this more, I might actually prefer if the type hints happened at the decoration stage, not the implementation stage:

```python
# don't love the __returns=...
# but I'm spitballing here
@ibis.udf.scalar.pyarrow(x=int, __returns=int)
def add_one(x: IntegerArray) -> IntegerArray:
    return pc.add(x, 1)

# switch to pandas implementation
# and the python type hints change (as they should)
# but the ibis hints don't change (as they shouldn't)

@ibis.udf.scalar.pandas(x=int, __returns=int)
def add_one(x: pd.Series) -> pd.Series:
    return x + 1

# can directly plug in 3rd party functions
# without needing to do a `def` block
from otherlib import process_series
ibis.udf.scalar.pandas(x=int, __returns=int)(process_series)

# or lambdas
# (is the name important to ibis???)
ibis.udf.scalar.pandas(x=int, __returns=int)(name="to_codes", lambda x: x.factorize[0])
```

**Comment by NickCrews** [comment]
I feel very strongly about this, is there anything I can do to move this forward before the next release (after which it will be harder to change the API)? Thanks so much, I appreciate the extra effort towards re-thinking about what you must imagine is an already-solved problem.

**Comment by cpcloud** [comment]
We are planning to release on Monday. UDF support is pretty new and is currently marked experimental, which means it's subject to breakage across non-major releases.

I don't think any changes we'd make to accommodate more mypy-accurate type hinting need to block the Monday release.

**Comment by cpcloud** [comment]
Another option might be

```
@ibis.udf.scalar.pyarrow(x=int, y=int, z=int).returns(float)
```

**Comment by cpcloud** [comment]
> (is the name important to ibis???)

Yes. We have to generate SQL for any calls to a UDF, so there needs to be name to use for the function call.

**Comment by NickCrews** [comment]
> @ibis.udf.scalar.pyarrow(x=int, y=int, z=int).returns(float)

I like that! Perhaps really embrace the builder pattern to be consistent between args and return types, to support possible per-implementor options such as [pyarrow's set_cpu_count()](https://arrow.apache.org/docs/python/api/misc.html), and to add a way to define the name separately if you use a lambda, or you want to override passed in callable's `__name__`. All of these are optional in the simple case.

```
@ibis.udf.scalar.pyarrow(n_cpus=2).consumes(x=int, y=int, z=int).returns(float).name("my_func")
def _private_name(...):
    ...
```

**Comment by NickCrews** [comment]
That fix looks awesome, thanks so much for the help here!
